pkg/kubelet/kubelet_pods.go

```diff
@@ -1521,6 +1522,18 @@ func getPhase(spec *v1.PodSpec, info []v1.ContainerStatus) v1.PodPhase {
 		// one container is running
 		return v1.PodRunning
 	case running == 0 && stopped > 0 && unknown == 0:
+		if utilfeature.DefaultFeatureGate.Enabled(features.PodDisruptionConditions) {
+			// Ensure all terminal pods, except for mirror and static pods, are transitioned to a terminal phase.
+			// The pod is deleted so its containers won't be restarted regardless of the restart policy.
+			if (pod.DeletionTimestamp != nil || podIsTerminal) && !kubetypes.IsMirrorPod(pod) && !kubetypes.IsStaticPod(pod) {
```

Comment:
The mirror pod / static pod thing is just a bug that could happen today (that has to get fixed).  So we'd remove those conditions before merge.

You mentioned that you were seeing pod container status missing?  I.e. without pod.DeletionTimestamp != nil, by the time SyncTerminatedPod is called there was a missing container status such that the pod is marked failed?  Can you link to a test run log that has that output?

Reply:
Yes, here is one such failure on infra, the log is in Artifacts: https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/115331/pull-kubernetes-node-e2e-containerd/1634122773481656320. 

I have one more reproduced locally with logging at 10. I will share it soon too.

Reply:
This is the kubelet log filtered by the pod name OR pod UID: https://files.fm/u/xf3nczy7w

Reply:
The same kubelet log but I added filtering by the container ID to trace it too: https://files.fm/u/bcvkj2k5t
We can see called:
`
I0310 15:24:12.981707  653463 kuberuntime_container.go:970] "Removing container" containerID="6bc4d88df30bc4bf9ec4b3bcda53d07514675f2d73827960003fd6467628e12e"
`

Reply:
I'm looking at https://github.com/kubernetes/kubernetes/pull/116482 as a possible cause, but there may be other reasons for static pod.

Reply:
> Can you try that change (should fix losing container status information to a GC). We should not need the pod.DeletionTimestamp check inside of getPhase then, and you can revert to just passing podSpec.

Pushed that change, running locally in a loop the test.

Reply:
The error rate is lower, but I got a failure, slightly different: https://files.fm/u/rqpry8dh8.

I think the issue is that the containers are not necessarily stopped inside `syncTerminatingPod`. Thus, it does not enter the check for the `isTerminalPod`. Then, the next invocation of getPhase is from `syncTerminatedPod`, but by then the container may already be lost. I think this scenario is rare, and possibly also existed when I was checking for `deletionTimestamp`, just didn't encounter it before.

Reply:
Right, this has always been possible.  However, since we have a kep now saying that we want to be precise, I think we might actually be able to fix it now by using the container status we fetch at the end of SyncTerminatingPod.  I'm pretty sure this scenario exists against master too (there is a test "should never report success for a pending container" that creates lots of pods and then deletes them quickly and I believe tolerates it).

In your latest run it *looks* like a GC happened, but GC should skip pods that are not fully terminated so I'll double check to make sure.

Reply:
> In your latest run it _looks_ like a GC happened, but GC should skip pods that are not fully terminated so I'll double check to make sure.

I got another failure with added logs and it was going via https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/pod_container_deletor.go.

Reply:
> However, since we have a kep now saying that we want to be precise

Well, it would be great to be precise, but not necessarily. In the KEP (for Jobs) what we care the most is that we get terminal phase eventually. Currently Job controller says `deletionTimestamp != nil`  is failure. One consequence of this is unnecessary pod restart in case of succeeded pods, but it is not terrible if they would continue to still occasionally happen.

Reply:
I think you should try calling

```
	apiPodStatus := kl.generateAPIPodStatus(pod, podStatus, true)
	if podStatusFn != nil {
		podStatusFn(&apiPodStatus)
	}
	kl.statusManager.SetPodStatus(pod, apiPodStatus)
```

after we fetch pod status at the end of the method - that should record the final state if we have it

Reply:
sg, pushed and testing locally

Reply:
I don't think we can address the static pods thing in this PR - I know why it's broken, but it's because admission uses the status manager as the local source of truth for whether a pod has been rejected, but we can't just bypass admission.

So in your code add a `if !kubetypes.IsStaticPod(pod)` around each of the blocks and add a TODO: to #116484 (that i'll track and own).

Reply:
Pushed, I'm wondering if this needs wrapping too. I assume not, because they are not on the API server: https://github.com/kubernetes/kubernetes/blob/9cba62087c2473d8739903a09ab5f8c4c6fa7caa/pkg/kubelet/status/status_manager.go#L876

Reply:
The test also fails for mirror pods: https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/115331/pull-kubernetes-node-e2e-containerd/1634319858520821760/

Reply:
I assume I can skip them too and TODO under the same issue.

Reply:
@smarterclayton after the recent changes there still exists a rare race condition, which is as follows:
1. `syncTerminatedPod` has all the containers terminated with exit code 0, is computes the status phase as `Succeeded` and updates the status via status_manager, which removes the status from Kubelet's cache:
https://github.com/kubernetes/kubernetes/blob/d34b0275a3b422f943340380a53069ba82bbf3bb/pkg/kubelet/status/status_manager.go#L835
2. `syncTerminatedPod` triggers `generateAPIPodStatus`, which does not see the recent cache status, as it is deleted from the previous step. Thus is fallbacks to the pod.Status: https://github.com/kubernetes/kubernetes/blob/d34b0275a3b422f943340380a53069ba82bbf3bb/pkg/kubelet/kubelet_pods.go#L1579 (which contains still the phase as `Running`). As the containers are removed in the meanwhile, `generateAPIPodStatus` generates `Failed` phase for the pod.
Issue: The phase is updated from `Succeeded` to  `Failed`.

I can see two solutions:
1. just make sure in the `status_manager.mergePodStatus` that we never transition a terminal phase. Implemented in the recent commit, but it still has a small issue that the containers themselves might be transitioned out of terminated (unless we additionally guard against that).
2. Delay deletion of the status from the pod status cache until `syncTerminatedPod` is executed. This could be done by passing an additional boolean via the `TerminatePod` function to indicate that the pod is terminated. It would need to be ORed with chacking for IsFinished() in pod_workers: https://github.com/kubernetes/kubernetes/blob/d34b0275a3b422f943340380a53069ba82bbf3bb/pkg/kubelet/pod_workers.go#L395. However, this by itself would only prevent `generateAPIPodStatus` to flip the phase, we would need to additionally adjust the logic in `kubelet_pods.convertStatusToAPIStatus` to make sure containers don't move out of terminated state in that case.

I prefer option 2. and I'm going to implement it unless other ideas.

Reply:
Here is the log with the example failure: https://files.fm/u/aqmnjkf89

Reply:
I've tested that with the current fix the tests are stable (though we may be getting containers in non-terminated state occasionally). I'm going to prototype option 2 to see if then we can fix the containers as well.

Reply:
Just pushed a commit aiming to delay cleaning of the cache until no further updates. Testing in a loop now.

Reply:
The recent change to delay pod deletion until SyncTerminatedPod appears to work, no flakes after 150 repeats. If so, I don't need the change in mergePodStatus, which would anyway not set the ExitCode properly.